Add Variational Autoencoder and GANs (#131)
johko/computer-vision-course: chapters/en/Unit 5 - Generative Models/gans.mdx

# Generative Adversarial Networks

> ## Section 2.1 Introduction to GANs

Author: I've resolved these grammatical errors. However, if there are any remaining please do let me know. Cheers :v:

> Generative Adversarial Networks (GANs) are a class of deep learning models introduced by [Ian Goodfellow](https://scholar.google.ca/citations?user=iYN86KEAAAAJ&hl=en) and his colleagues in 2014. The core idea behind GANs is to train a generator network to produce data that is indistinguishable from real data, while simultaneously training a discriminator network to differentiate between real and generated data.

Reviewer: Sometimes the `discriminator` is refered to as the `critic`. I have always liked this term better as this directly comes from the analogy of a painter (generator) and an art critic (discriminator). This analogy was part of the [GAN course from DeepLearning.ai](https://www.deeplearning.ai/courses/generative-adversarial-networks-gans-specialization/).

This is not a suggestion. Might be interesting to pursue this analogy in the introductory paragraph. Really drives the intuition home.

Author: Added

* **Image Generation:**
    - **GANs:**
        * **Strengths:** Generate higher quality images, especially for complex data with sharp details and realistic textures.
        * **Weaknesses:** Can be more difficult to train and prone to instability.
        * **Example:** A GAN-generated image of a bedroom is likely to be indistinguishable from a real one, while a VAE-generated bedroom might appear blurry or have unrealistic lighting.
        IMAGE
    - **VAEs:**
        * **Strengths:** Easier to train and more stable than GANs
        * **Weaknesses:** May generate blurry, less detailed images with unrealistic features.
* **Other Tasks:**
    - **GANs:**
        * **Strengths:** Can be used for tasks like super-resolution and image-to-image translation.
        * **Weaknesses:** May not be the best choice for tasks that require a smooth transition between data points
    - **VAEs:**
        * **Strengths:** Widely used for tasks like image denoising and anamoly detection.

>         * **Weaknesses:** May not be as effective as GANs for tasks that require high-quality image generation.

Reviewer: I would always cite different papers to prove the pointers mentioned. If not papers, reports of experiments conducted.

Author: I got these points after some Google search and from some medium posts. I don't know of any papers which directly reference a comparison like this between GANs and VAEs so if you know of any paper do let me know and I will add it. Cheers :v:

Author: I think this should work. Although I'm not sure about the sources
* [What are the fundamental differences between VAE and GAN for image generation?](https://ai.stackexchange.com/questions/25601/what-are-the-fundamental-differences-between-vae-and-gan-for-image-generation)
* [Issues with GAN and VAE models](https://stats.stackexchange.com/questions/541775/issues-with-gan-and-vae-models)
* [VAE Vs. GAN For Image Generation](https://www.baeldung.com/cs/vae-vs-gan-image-generation)
* [Diffusion Models vs. GANs vs. VAEs: Comparison of Deep Generative Models](https://towardsai.net/p/machine-learning/diffusion-models-vs-gans-vs-vaes-comparison-of-deep-generative-models)

Here's a table summarizing the key differences:
|Feature|GANs|VAEs|
|-------|-----|---|
|Image Quality|Higher|Lower|
|Ease of Training|More difficult|Easier|
|Stability|Less Stable|More Stable|

> |Applications|Image Generation, Super-resolution, image-to-image translation|Image Denoising, Anamoly Detection, Signal Analysis|

Reviewer: This is so neat! A summarizer of all the pointers is a very thoughtful.

Ultimately, the best choice depends on one's specific needs and priorities. If one need high-quality images for tasks like generating realistic faces or landscapes, then a GAN might be the better choice. However, if one needs a model that is easier to train and more stable, then a VAE might be a better option.

> ## Section 2.3 Training GANs

Reviewer: [Slide 12: Using Math to Communicate](https://docs.google.com/presentation/d/e/2PACX-1vRau7IXhvdKKOZQt3hujyw5dNltExsmxfHTMbHTpwUXK4mst-lXMPkLQL8MRNsVVONWRgBdTf5W3k4c/pub?start=false&loop=false&delayms=3000&slide=id.g157b6906e85_0_25)
[Slide 13: The Samplings](https://docs.google.com/presentation/d/e/2PACX-1vRau7IXhvdKKOZQt3hujyw5dNltExsmxfHTMbHTpwUXK4mst-lXMPkLQL8MRNsVVONWRgBdTf5W3k4c/pub?start=false&loop=false&delayms=3000&slide=id.g157b6906e85_0_38)
[Slide 14: Discriminator](https://docs.google.com/presentation/d/e/2PACX-1vRau7IXhvdKKOZQt3hujyw5dNltExsmxfHTMbHTpwUXK4mst-lXMPkLQL8MRNsVVONWRgBdTf5W3k4c/pub?start=false&loop=false&delayms=3000&slide=id.g157b6906e85_0_50)
[Slide 15: Generator](https://docs.google.com/presentation/d/e/2PACX-1vRau7IXhvdKKOZQt3hujyw5dNltExsmxfHTMbHTpwUXK4mst-lXMPkLQL8MRNsVVONWRgBdTf5W3k4c/pub?start=false&loop=false&delayms=3000&slide=id.g157bb6d6115_0_2)
[Slide 16: Revisited](https://docs.google.com/presentation/d/e/2PACX-1vRau7IXhvdKKOZQt3hujyw5dNltExsmxfHTMbHTpwUXK4mst-lXMPkLQL8MRNsVVONWRgBdTf5W3k4c/pub?start=false&loop=false&delayms=3000&slide=id.g157bb6d6115_0_19)

These might help you build a better intuition about the min max game.

A [desmos](https://www.desmos.com/calculator/nw6m38xtij) demo for the mathametical equation also helps here.

# Generative Adversarial Networks

## Introduction
Generative Adversarial Networks (GANs) are a class of deep learning models introduced by [Ian Goodfellow](https://scholar.google.ca/citations?user=iYN86KEAAAAJ&hl=en) and his colleagues in 2014. The core idea behind GANs is to train a generator network to produce data that is indistinguishable from real data, while simultaneously training a discriminator network to differentiate between real and generated data.
* **Architecture overview:** GANs consist of two main components: `the generator` and `the discriminator`
* **Generator:** The generator takes random noise $$z$$ as input and generates synthetic data samples. Its goal is to create data that is realistic enough to deceive the discriminator.
* **Discriminator:** The discriminator, akin to a detective, evaluates whether a given sample is real (from the actual dataset) or fake (generated by the generator). Its objective is to become increasingly accurate in distinguishing between real and generated samples.

> A common analogy which can be found online is that of an art forger/painter (the geneator) which tries to forge paintings and an art investigator/critic (the discriminator) which triest to detect imitations

Reviewer 2: ```suggestion
A common analogy that can be found online is that of an art forger/painter (the generator) which tries to forge paintings and an art investigator/critic (the discriminator) who tries to detect imitations.
```

![Lilian Weng GAN Figure](https://lilianweng.github.io/posts/2017-08-20-gan/GAN.png)

> ## Section 2.2 GANs vs VAEs

Reviewer 2: ```suggestion
## GANs vs VAEs
```
No need for numbering

> GANs and VAEs are both popular generative models in the field of machine learning, but they have different strengths and weaknesses. Whether one is "better" than the other depends on the specific task and requirements. Here's a breakdown of their strengths and weaknesses.

Reviewer 2: ```suggestion
GANs and VAEs are both popular generative models in machine learning, but they have different strengths and weaknesses. Whether one is "better" depends on the specific task and requirements. Here's a breakdown of their strengths and weaknesses.
```

* **Image Generation:**
    - **GANs:**

>         * **Strengths:** Generate higher quality images, especially for complex data with sharp details and realistic textures.

Author: I've addressed the other issues but I believe removing the indentations might impact the structure of the content. Please let me know whether it should still be removed or kept.

        * **Weaknesses:** Can be more difficult to train and prone to instability.
        * **Example:** A GAN-generated image of a bedroom is likely to be indistinguishable from a real one, while a VAE-generated bedroom might appear blurry or have unrealistic lighting.
        ![Example of GAN-Generated bedrooms taken from Unsupervised Representation Learning with Deep Convolutional Generative Adversarial Networks, 2015](https://machinelearningmastery.com/wp-content/uploads/2019/04/Example-of-GAN-Generated-Photographs-of-Bedrooms.png)
    - **VAEs:**
        * **Strengths:** Easier to train and more stable than GANs
        * **Weaknesses:** May generate blurry, less detailed images with unrealistic features.
* **Other Tasks:**
    - **GANs:**
        * **Strengths:** Can be used for tasks like super-resolution and image-to-image translation.
        * **Weaknesses:** May not be the best choice for tasks that require a smooth transition between data points
    - **VAEs:**

>         * **Strengths:** Widely used for tasks like image denoising and anamoly detection.

Reviewer 2: ```suggestion
        * **Strengths:** Widely used for tasks like image denoising and anomaly detection.
```

        * **Weaknesses:** May not be as effective as GANs for tasks that require high-quality image generation.

Here's a table summarizing the key differences:
|Feature|GANs|VAEs|
|-------|-----|---|
|Image Quality|Higher|Lower|
|Ease of Training|More difficult|Easier|
|Stability|Less Stable|More Stable|
|Applications|Image Generation, Super-resolution, image-to-image translation|Image Denoising, Anamoly Detection, Signal Analysis|

> Ultimately, the best choice depends on one's specific needs and priorities. If one need high-quality images for tasks like generating realistic faces or landscapes, then a GAN might be the better choice. However, if one needs a model that is easier to train and more stable, then a VAE might be a better option.

Reviewer 2: ```suggestion
Ultimately, the best choice depends on one's specific needs and priorities. If one needs high-quality images for tasks like generating realistic faces or landscapes, then a GAN might be the better choice. However, if one needs a model that is easier to train and more stable, then a VAE might be a better option.
```
